dev/benchmark/sysbench-in-k8s.md

```diff
@@ -0,0 +1,452 @@
+---
+title: TiDB in Kubernetes Sysbench 性能测试
+category: benchmark
+---
+
+# TiDB in Kubernetes Sysbench 性能测试
+
+随着 [TiDB Operator GA 发布](https://pingcap.com/blog-cn/tidb-operator-1.0-ga/)，越来越多用户开始使用 TiDB Operator 在 Kubernetes 中部署管理 TiDB 集群。在本次测试中，我们选择 GKE 平台做了一次深入、全方位的测试，方便大家了解 TiDB 在 Kubernetes 中性能影响因素。
+
+## 目的
+
+- 测试典型公有云平台上 TiDB 性能数据
+- 测试公有云平台磁盘、网络、CPU 以及不同 Pod 网络下对 TiDB 性能的影响
+
+## 环境
+
+### 版本与配置
+
+本次测试统一使用 TiDB v3.0.1 版本进行测试。
+
+TiDB Operator 使用 v1.0.0 版本。
+
+PD、TiDB 和 TiKV 实例数均为 3 个。各组件分别作了如下配置，未配置部分使用默认值。
+
+PD:
+
+```toml
+[log]
+level = "info"
+[replication]
+location-labels = ["region", "zone", "rack", "host"]
+```
+
+TiDB:
+
+```toml
+[log]
+level = "error"
+[prepared-plan-cache]
+enabled = true
+[tikv-client]
+max-batch-wait-time = 2000000
+```
+
+TiKV:
+
+```toml
+log-level = "error"
+[server]
+status-addr = "0.0.0.0:20180"
+grpc-concurrency = 6
+[readpool.storage]
+normal-concurrency = 10
+[rocksdb.defaultcf]
+block-cache-size = "14GB"
+[rocksdb.writecf]
+block-cache-size = "8GB"
+[rocksdb.lockcf]
+block-cache-size = "1GB"
+[raftstore]
+apply-pool-size = 3
+store-pool-size = 3
+```
+
+### TiDB 数据库参数配置
+
+```
+set global tidb_hashagg_final_concurrency=1;
+set global tidb_hashagg_partial_concurrency=1;
+set global tidb_disable_txn_auto_retry=0;
+```
+
+### 硬件配置
+
+#### 机器型号
+
+在单可用区测试中，我们选择了如下型号机器：
+
+| 组件     | 实例类型       | 数量  |
+| :---     | :---------     | :---- |
+| PD       | n1-standard-4  | 3     |
+| TiKV     | c2-standard-16 | 3     |
+| TiDB     | c2-standard-16 | 3     |
+| Sysbench | c2-standard-30 | 1     |
+
+在多可用区相对单可用区性能对比中，因为测试时 (2019.08) GCP 的 Region 均没有 3 个可用区可同时提供 c2 机器，我们选择了如下机器型号进行测试：
+
+| 组件     | 实例类型       | 数量  |
+| :---     | :---------     | :---- |
+| PD       | n1-standard-4  | 3     |
+| TiKV     | n1-standard-16 | 3     |
+| TiDB     | n1-standard-16 | 3     |
+| Sysbench | n1-standard-16 | 3     |
+
+在高并发读测试中，压测端 sysbench 对 CPU 需求较高，需选择多核较高配置型号，避免压测端成为瓶颈。
+
+> **注意：**
+>
+> GCP 不同的区域可用机型不同。同时测试中发现磁盘性能表现也存在差异，我们统一在 us-central1 中申请机器测试。
+
+#### 磁盘
+
+GKE 当前的 NVMe 磁盘还在 Alpha 阶段，使用需特殊申请，不具备普遍意义。测试中，本地 SSD 盘统一使用 iSCSI 接口类型。挂载参数参考[官方建议](https://cloud.google.com/compute/docs/disks/performance#optimize_local_ssd)增加了 `discard,nobarrier` 选项。完整挂载例子如下：
+
+```
+sudo mount -o defaults,nodelalloc,noatime,discard,nobarrier /dev/[LOCAL_SSD_ID] /mnt/disks/[MNT_DIR]
+```
+
+#### 网络
+
+GKE 网络模式使用具备更好扩展性以及功能强大的 [VPC-Native](https://cloud.google.com/kubernetes-engine/docs/how-to/alias-ips) 模式。在性能对比中，我们分别对 TiDB 使用 Kubernetes Pod 和 Host 网络分别做了测试。
+
+#### CPU
+
+在单可用集群测试中，我们为 TiDB/TiKV 选择 c2-standard-16 机型测试。在单可用与多可用集群的对比测试中，因为 GCP 没有区域可以同时在 3 个可用区里申请 c2-standard-16 机型。我们选择了 n1-standard-16 机型测试。
+
+### 操作系统及参数
+
+GKE 支持两种操作系统：COS (Container Optimized OS) 和 Ubuntu 。在 Point Select 测试中进行了对比测试。其余测试中，统一使用 Ubuntu 系统进行测试。
+
+内核统一做了如下配置:
+
+```shell
+sysctl net.core.somaxconn=32768
```

Comment:
The net.core.somaxconn in the container is still 128 even if we set 32768 on the host, will this have any impact on the test?

Reply:
this sysctl parameter determines [backlog of the listen function](https://golang.org/src/net/sock_linux.go), however, we bench performance after all TCP connections are established, so it has no impact on our perf results